crates/qasm2/src/lex.rs

```diff
@@ -0,0 +1,772 @@
+            b'/' => {
+                if let Some(b'/') = self.peek_byte()? {
+                    self.advance_line()?;
+                    return self.next(context);
+                } else {
+                    TokenType::Slash
+                }
+            }
+            b'-' => {
+                if let Ok(Some(b'>')) = self.peek_byte() {
+                    self.col += 1;
+                    TokenType::Arrow
+                } else {
+                    TokenType::Minus
+                }
+            }
+            b'=' => {
+                if let Ok(Some(b'=')) = self.peek_byte() {
+                    self.col += 1;
+                    TokenType::Equals
+                } else {
+                    return Err(QASM2ParseError::new_err(
+                        "single equals '=' is never valid".to_owned(),
+                    ));
+                }
+            }
+            b'0'..=b'9' | b'.' => self.lex_numeric(start_col)?,
+            b'a'..=b'z' | b'A'..=b'Z' => self.lex_textlike(start_col)?,
+            c @ (b'"' | b'\'') => {
+                if self.strict && c != b'"' {
+                    return Err(QASM2ParseError::new_err(message_generic(
+                        Some(&Position::new(&self.filename, self.line, start_col)),
+                        "[strict] paths must be in double quotes (\"\")",
+                    )));
+                } else {
+                    self.lex_filename(c, start_col)?
+                }
+            }
+            c => {
+                return Err(QASM2ParseError::new_err(message_generic(
+                    Some(&Position::new(&self.filename, self.line, start_col)),
+                    &format!(
+                        "encountered '{}', which doesn't match any valid tokens",
+                        // Non-ASCII bytes should already have been rejected by `next_byte()`.
+                        c as char,
+                    ),
+                )));
+            }
+        };
+        Ok(Some(Token {
+            ttype,
+            line: self.line,
+            col: start_col,
+            index: if ttype.variable_text() {
+                context.index(&self.line_buffer[start_col..self.col])
+            } else {
+                usize::MAX
+            },
+        }))
+    }
+
+    /// Get an optional reference to the next token in the iterator stream without consuming it.
+    /// This is a direct analogue of the same method on the [std::iter::Peekable] struct, except it
+    /// is manually defined here to avoid hiding the rest of the public fields of the [TokenStream]
+    /// struct itself.
+    pub fn peek(&mut self, context: &mut TokenContext) -> PyResult<Option<&Token>> {
+        if self.peeked.is_none() {
+            self.peeked = Some(self.next_inner(context)?);
+        }
+        Ok(self.peeked.as_ref().unwrap().as_ref())
+    }
+
+    pub fn next(&mut self, context: &mut TokenContext) -> PyResult<Option<Token>> {
+        match self.peeked.take() {
```

Comment:
Chef's kiss! 💋

Reply:
`Option::take` is definitely nice, yeah!